src/universal/utils/swarm/MediaSwarm.ts

```diff
@@ -0,0 +1,213 @@
+}
+
+const qualityDict: LocalQualityDict = {low: null, med: null, high: null}
+
+interface Config extends SwarmConfig {
+  dispatchState: Dispatch<ReducerAction<typeof reducerSwarm>>
+  trebuchet: Trebuchet
+  userId: string
+}
+
+const qualityConstraints = {
+  audioOnly: {audio: true, video: false},
+  low: {audio: true, video: {width: 64, height: 64}},
+  lowVideo: {audio: false, video: {width: 64, height: 64}}
+}
+
+export default class MediaSwarm extends FastRTCSwarm {
+  trebuchet: Trebuchet
+  dispatchState: Dispatch<ReducerAction<typeof reducerSwarm>>
+  userId: string
+  localStreams = {
+    cam: {...qualityDict},
+    screen: {...qualityDict}
+  }
+
+  constructor (config: Config) {
+    super(config)
+    this.dispatchState = config.dispatchState
+    this.trebuchet = config.trebuchet
+    this.userId = config.userId
+    this.trebuchet.on(Events.DATA, this.handleSignal)
+    this.on('signal', (signal) => {
+      this.trebuchet.send(JSON.stringify({type: 'WRTC_SIGNAL', signal}))
+    })
+    this.on('stream', this.handleStream)
+    this.on('close', this.handleClose)
+    // see if we fixed all the signaling errors
+    this.on('error', console.error)
+
+    this.on('connection', console.log)
+    this.on('open', (peer) => {
+      console.log('open', peer)
+    })
+  }
+
+  private handleClose = () => {
+    this.dispatchState({type: 'removeStream', userId: this.userId})
+  }
+
+  private handleSignal = (data) => {
+    if (typeof data !== 'string') return
+    try {
+      const payload = JSON.parse(data)
+      if (payload.type === 'WRTC_SIGNAL') {
+        this.dispatch(payload.signal)
+      }
+    } catch (e) {
+      /**/
+    }
+  }
+
+  private handleStream: FastRTCSwarmEvents['stream'] = (stream, streamName, peer) => {
+    if (streamName !== 'cam') {
+      console.log('screen sharing not yet implemented')
+      return
+    }
+    console.log('got stream')
+    const userId = peer.userId!
+    // start out by setting hasVideo to false in case the peer is muting their video feed due to the chrome bug
+    const streamUI = {
+      hasAudio: false,
+      hasVideo: false,
+      isAudioBlocked: false,
+      isVideoBlocked: false
+    } as StreamUI
+    this.dispatchState({type: 'setStream', streamName, userId, streamUI})
+
+    const setStreamState = (streamUI: Partial<StreamUI>) => {
+      this.dispatchState({type: 'setStream', streamName, userId, streamUI})
+    }
+
+    stream.getTracks().forEach((track) => {
+      const field = track.kind === 'video' ? 'hasVideo' : 'hasAudio'
+      const video = document.createElement('video')
+
+      // chrome workaround for not detecting empty tracks
+      video.srcObject = new MediaStream([track])
+      video.onloadedmetadata = () => {
+        console.log('METADATA LOADED (chrome unmuted workaround)')
+        if (track.readyState === 'live') {
+          setStreamState({[field]: true})
+        }
+      }
+
+      track.onended = () => {
+        console.log('ENDED')
```

Comment:
You want to keep these console log events?

Reply:
for alpha definitely. it'll help with sentry errors & debugging